### ISSUE TITLE
memsync - support for calculation of 'l' size parameter?

### ISSUE THREAD

**Issue by jtlz2** [pleiszenburg/zugbruecke#33]
I have a `memsync` as follows:    

```
func.memsync = [
        {
                'p': [0,'data'],
                'l: **???**,
                '_t': ctypes.c_ubyte
        },
        {
                'p': [1,'data'],
                'l': [1,'size'],
                '_t': ctypes.c_ubyte
        }
        ]
```

However, argument 0 has no `size` parameter. I cannot add one because the C prototype is not defined by me. It is possible to derive a `size` from two other parameters in the `struct`s for argument 0.

Does `memsync` support calculation of 'l' in python? If not, is there any chance you could add such functionality?

Thanks as ever!

**Comment by s-m-e** [comment]
I can certainly add such a functionality. A typical use case is images with a width and height specified separately. The [function for packing the memory](https://github.com/pleiszenburg/zugbruecke/blob/8652a873027348c1cb25f92bc1ca97fd09a9be6d/src/zugbruecke/core/data/memory.py#L62) before it is shipped to the other side was next on my list for a rewrite anyway - it's fundamentally flawed.

**Comment by jtlz2** [comment]
@s-m-e That would be great - thanks so much!

**Comment by s-m-e** [comment]
@jtlz2 I pushed a new (experimental) branch named `memsync_lambda`. Have a look at [this test code](https://github.com/pleiszenburg/zugbruecke/blob/1fdb878ce667f2d6753e516b0cb739d35a3061c7/tests/test_bubblesort_segments.py#L56). It's pretty much self-explanatory I hope.

**Comment by jtlz2** [comment]
@s-m-e Brilliant - thank you! Is `_f` an optional parameter, or must I include it in every `memsync` dictionary? Does the new syntax for `_l` replace the old syntax for every variable there?

**Comment by jtlz2** [comment]
In the example,
```
			{
				'p': [0], # "path" to argument containing the pointer
				'l': ([1], [2]), # "path" to arguments containing information on length
				'_f': lambda x, y: x * y, # function computing length from relevant arguments
				'_t': ctypes.c_float # type of argument (optional, default char/byte): sizeof(type) * length == bytes
}
```

I am not sure I have the path for `'l'` right - if `x` and `y` are in positions 1 and 2 in the list of `_fields_` for the Struct, should 'l' = ([1], [2])? Is the `lambda` then just as is?

The lambda is really nice by the way because it can be set to `lambda x: x`, `lambda x:const` etc.

**Comment by jtlz2** [comment]
Apologies - am confused:

Should I have:

```
                'p': [1,'data'],
                'l': ([1,'height'],[1,'width']),
                '_f': lambda x, y: x * y, # function computing length from relevant arguments
                '_t': ctypes.c_ubyte
```

I don't have the syntax quite right yet....

**Comment by s-m-e** [comment]
@jtlz2 Sry, I did not bother to write documentation for it yesterday - lack of time. First, I usually try to not break backwards compatibility. `_f` is optional. Old `memsync` statements without it are not affected. If (and only if) you specify `_f` (and fill it with a function), the `l` (length path) parameter has to become a tuple of path lists (instead of one single path list). The parser then tries to extract the numbers from the elements the length paths are pointing to, for example an integer for `width` and another integer for `height`, and uses them as parameters for the function specified in `_f`. `_f` does not need to be a lambda function. You can also plug in a real Python function if you want. The basic (only) requirement is that the number of path lists in the tuple has to be identical to the number of arguments the function accepts.

Your example with `width` and `height` looks about right.

**Comment by jtlz2** [comment]
@s-m-e Thanks for the info - I think I have it right now (for argument `1`):

```
{
'p': [1,'data'],
 'l': ([1,1],[1,2]),
'_f': lambda x, y: x * y, # function computing length from relevant arguments
'_t': ctypes.c_ubyte
}
```

`[1,1]` and `[1,2]` refer respectively to the width and height parameters.

**Comment by s-m-e** [comment]
@jtlz2 I am not sure this will work (at least it might not cause an error which it should): You have a struct with a field named ``data``. It's the 2nd argument of your function (i.e. the leading ``1`` in the paths). The same struct also contains fields representing information about the length of ``data``. Structs do not have indices, only named fields - to the tailing ``1`` and ``2`` are wrong from my perspective. You'd actually have to use the names of the fields as strings here. Your initial definition with ``'l': ([1,'height'],[1,'width'])`` was right.

**Comment by s-m-e** [comment]
Feature released in v0.0.10.
